### PR TITLE
Adds usage to picdarExport tool

### PR DESCRIPTION
**Pull request by kenoir**
This changeset starts adding functionality to the picdar export tool in
order to additionally import usages.

Requires ES mapping update before release.

### CODE REVIEW

**Comment by kenoir** [comment]
**TODO:**
- [ ] add optional source
- [ ] add notes (not orderedBy!)
- [ ] optional size
- [ ] optional layout id
- [ ] section id = section name
- [ ] orderedBy = Picdar
- [ ] Add usage "marked as sent"

File: common-lib/src/main/scala/com/gu/mediaservice/lib/argo/model/Action.scala [hunk 1]
```diff
@@ -10,11 +10,9 @@ import play.api.libs.functional.syntax._
...
 
 object Action {
-
   implicit val actionWrites: Writes[Action] = (
     (__ \ "name").write[String] ~
-      (__ \ "href").write[String].contramap((_: URI).toString) ~
-      (__ \ "method").write[String]
-    )(unlift(Action.unapply))
-
+    (__ \ "href").write[String].contramap((_: URI).toString) ~
```
Comment (jamesgorrie) [inline common-lib/src/main/scala/com/gu/mediaservice/lib/argo/model/Action.scala]: We could write an implicit that does this for URIs everywhere (I've seen it else where)...

File: common-lib/src/main/scala/com/gu/mediaservice/lib/aws/DynamoDB.scala [hunk 2]
```diff
@@ -29,6 +29,13 @@ class DynamoDB(credentials: AWSCredentials, region: Region, tableName: String) {
 
   val IdKey = "id"
 
+  def exists(id: String)(implicit ex: ExecutionContext): Future[Boolean] = Future {
+      table.getItem(new GetItemSpec().withPrimaryKey(IdKey, id))
+  } map(result => Option(result) match {
```
Comment (theefer) [inline common-lib/src/main/scala/com/gu/mediaservice/lib/aws/DynamoDB.scala]: [`isDefined`](http://www.scala-lang.org/api/current/index.html#scala.Option@isDefined:Boolean)?

File: common-lib/src/main/scala/com/gu/mediaservice/model/usage/UsageStatus.scala [hunk 3]
```diff
@@ -11,15 +11,18 @@ trait UsageStatus {
...
 object UsageStatus {
+  def apply(status: String): UsageStatus = status match {
+    case "pending" => PendingUsageStatus()
+    case "published" => PublishedUsageStatus()
+  }
+
   implicit val reads: Reads[UsageStatus] = JsPath.read[String].map {
     case "pending" => PendingUsageStatus()
     case "published" => PublishedUsageStatus()
   }
```
Comment (theefer) [inline common-lib/src/main/scala/com/gu/mediaservice/model/usage/UsageStatus.scala]: Could reuse `apply` above then?

File: project/Build.scala [hunk 4]
```diff
@@ -17,6 +17,7 @@ object Build extends Build {
       scalaVersion := "2.11.6",
       scalaVersion in ThisBuild := "2.11.6",
       organization := "com.gu",
+      fork in run := true,
```
Comment (theefer) [inline project/Build.scala]: Ooh what's this?

File: picdar-export/src/main/scala/com.gu.mediaservice.picdarexport/lib/usage/ResponseParser.scala [hunk 5]
```diff
@@ -0,0 +1,17 @@
+package com.gu.mediaservice.picdarexport.lib.usage
+
+
+object ResponseParser {
```
Comment (theefer) [inline picdar-export/src/main/scala/com.gu.mediaservice.picdarexport/lib/usage/ResponseParser.scala]: Could be useful to dump a sample response in a comment here for future reference?

File: picdar-export/src/main/scala/com.gu.mediaservice.picdarexport/lib/usage/PrintUsageRequestFactory.scala [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+object PrintUsageRequestFactory {
+  def create(picdarUsageRecords: List[PicdarUsageRecord], mediaId: String): Option[PrintUsageRequest] = {
+    val printUsageRecords = picdarUsageRecords.zipWithIndex.map {
+      case (record: PicdarUsageRecord, i: Int) => {
+        val metadata = PrintUsageMetadata(
+          sectionName = record.sectionName,
+          issueDate = record.publicationDate,
+          pageNumber = record.page,
+          storyName = record.productionName,
+          publicationCode = "XXX",
```
Comment (theefer) [inline picdar-export/src/main/scala/com.gu.mediaservice.picdarexport/lib/usage/PrintUsageRequestFactory.scala]: :eyes: :question:

File: picdar-export/src/main/scala/com.gu.mediaservice.picdarexport/lib/picdar/UsageApi.scala [hunk 7]
```diff
@@ -0,0 +1,26 @@
...
+  import com.gu.mediaservice.picdarexport.lib.Config.picdarUsageApiUrl
+
+  def get(urn: String): Future[List[PicdarUsageRecord]] = Future {
+    val response = Http(s"$picdarUsageApiUrl").param("xurn",urn).asString
+    val responseString = response.body.trim
+
+    val responseBody = responseString.isEmpty match {
+      case true => None
+      case false => Some(responseString)
+    }
```
Comment (theefer) [inline picdar-export/src/main/scala/com.gu.mediaservice.picdarexport/lib/picdar/UsageApi.scala]: Alternatively, you can wrap it in an `Option` and then use `filter(_.nonEmpty)`

File: picdar-export/src/main/scala/com.gu.mediaservice.picdarexport/lib/db/ExportDynamoDB.scala [hunk 8]
```diff
@@ -154,17 +183,61 @@ class ExportDynamoDB(credentials: AWSCredentials, region: Region, tableName: Str
...
+      pages <- query
+      items <- pages
+    } yield items
+
+    items.map(AssetRef(_))
+  }
+
+  // Type this method so we can coerce items into whatever you want!
+  import com.amazonaws.services.dynamodbv2.document.Item
+  def getUrnsForNotFilledFields[T](dateRange: DateRange, attrs: Set[String])(f: Item => T): Future[Seq[T]] = {
```
Comment (theefer) [inline picdar-export/src/main/scala/com.gu.mediaservice.picdarexport/lib/db/ExportDynamoDB.scala]: Could this be used to optimise the existing massive scan operations?

**Comment by theefer** [comment]
A few comments, but looks really nice!

File: project/Build.scala (hunk 4 above)
Comment (kenoir) [inline project/Build.scala]: http://www.scala-sbt.org/0.13/docs/Forking.html

Before this change `System.exit` was terminating the JVM process `sbt` was using so needed to be restart after each run. Asking `sbt` to fork a new JVM when performing `run` commands means that when that process exits we return to `sbt` ready to run another command.

File: picdar-export/src/main/scala/com.gu.mediaservice.picdarexport/lib/usage/ResponseParser.scala (hunk 5 above)
Comment (kenoir) [inline picdar-export/src/main/scala/com.gu.mediaservice.picdarexport/lib/usage/ResponseParser.scala]: Good idea.

File: picdar-export/src/main/scala/com.gu.mediaservice.picdarexport/lib/db/ExportDynamoDB.scala (hunk 8 above)
Comment (kenoir) [inline picdar-export/src/main/scala/com.gu.mediaservice.picdarexport/lib/db/ExportDynamoDB.scala]: I almost did it here, but I feel like it's probably too big a change for this PR. I'd like to try this out in PROD a bit then make the change when we're confident this approach performs well.

File: project/Build.scala (hunk 4 above)
Comment (theefer) [inline project/Build.scala]: Ah cool, thanks! :+1:

File: picdar-export/src/main/scala/com.gu.mediaservice.picdarexport/lib/db/ExportDynamoDB.scala (hunk 8 above)
Comment (theefer) [inline picdar-export/src/main/scala/com.gu.mediaservice.picdarexport/lib/db/ExportDynamoDB.scala]: Completely agree to keeping the change separate :+1:

**Comment by kenoir** [comment]
@akash1810 so something snapped and I think you are right about squashing feature branches so i'm going to do it from now on -> https://github.com/guardian/grid/pull/1648

**Pull request closed without merging**